RFC-0023-Unified-Memory-for-Pytorch.md

```diff
@@ -0,0 +1,127 @@
+
+The default value for enabled_uvm will be False. It is expected that when setting enabled_uvm to True, that it happens early in the Pytorch program before any allocations have taken place and then left on throughout. In order to enforce this, there is no proposed set_disabled_uvm() call. 
+
+`torch.cuda.set_uvm_move_on_copy()`
+
+This API will also set a flag (`enabled_uvm_move_on_copy`) in the globalContext. This flag will determine if tensor`.to()` should move the tensor to the new device instead of creating a copy. Aside from this flag, two other conditions must be met for a move to occur. 
+
+1. UVM is enabled 
+2. The `.to()` only changes the device and does not change any other tensor attribute. 
+
+`Torch.cuda.uvm_to(Tensor t)`
+
+This API is a new alternate “to” function which moves a UVM tensor from one device to another. 
+
+With the default tensor`.to()` function, a copy of the tensor is returned with the requested changes.  This is inefficient in UVM mode if the only desired change is to move a tensor from one device to another.  UVM will manage the migration of the tensor as needed. Allocating more storage via a call to `at::empty` isn’t needed. This can also be accomplished by setting the `torch.cuda.set_uvm_move_on_copy(True)` and then using tensor`.to()` and only changing the device, however there may be reasons to keep tensor`.to()` untouched and to use this function to target specific areas for copy elimination. 
+
+### C++
+Additionally, we propose to add a check on the environment for the `PYTORCH_ENABLE_UNIFIED_MEMORY` variable. When set, the same global context enablement flag will be set. 
+
+## Backend Changes
+To enable usage of the managed memory APIs in CUDA/ROCm, minimal backend changes are proposed below and focus on the memory allocator(s), empty tensors and copies. 
+
+### Memory Allocation
+Currently there are different allocators used for CPU (CPUAllocator) , pinned host memory (CachingHostAllocator), and device memory (CUDACachingAllocator) memory allocations.  
+
+ - When UVM is enabled, we propose forgoing all traditional allocators and working with a newly developed CachingManagedAllocator. 
+ - This will provide benefits in a few areas. We can continue to develop and improve on Unified memory allocations without the risk of introducing performance degradation, or errors to the existing CUDACachingAllocator. In addition, we expect the behavior of Managed memory to be different than that of traditional discrete GPU/CPU memory. For example, managed memory is available to all GPUs on all streams by default and can move between host and any number of GPUs at will. This will provide intricacies to the caching methodology that may conflict with the existing caching allocators. 
```

Comment:
about the stream part - the reason caching allocator has to keep track of streams for deallocation is to make deallocation synchronization free and be able to reuse the same buffer on another stream (after injecting a pair of events). I'd imagine you need to do the same here too

but in general agree on having separate allocator and being able to customize behavior